src/server/common/componentUtils.ts

```diff
@@ -0,0 +1,209 @@
+/**
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+type Operator = '<truthy>'|'='|'*='|'|='|'^='|'$='|'~=';
+export type ParsedComponentAttribute = {
+  jsonPath: string[],
+  op: Operator,
+  value: any,
+  caseSensetive: boolean,
+};
+
+export type ParsedComponentSelector = {
+  name: string,
+  attributes: ParsedComponentAttribute[],
+};
+
+export function checkComponentAttribute(obj: any, attr: ParsedComponentAttribute) {
+  for (const token of attr.jsonPath) {
+    if (obj)
```

Comment:
Let's check `typeof obj === 'object'`? We can also throw a nice error for this, like "foo.bar.baz is not an object". Are should we follow `a?.b?.c` pattern here, perhaps we already do?

Reply:
a?.b?.c is the actual logic.